Source/FieldSolver/WarpXPushFieldsHybridPIC.cpp

```diff
@@ -0,0 +1,254 @@
+        FillBoundaryB(guard_cells.ng_FieldSolver, WarpX::sync_nodal_points);
+    }
+
+    // Extrapolate the ion current density to t=n+1 using
+    // J_i^{n+1} = 1/2 * J_i^{n-1/2} + 3/2 * J_i^{n+1/2}, and recalling that
+    // now current_fp_temp = J_i^{n} = 1/2 * (J_i^{n-1/2} + J_i^{n+1/2})
+    for (int lev = 0; lev <= finest_level; ++lev)
+    {
+        for (int idim = 0; idim < 3; ++idim) {
+            MultiFab::LinComb(
+                *current_fp_temp[lev][idim],
+                -1._rt, *current_fp_temp[lev][idim], 0,
+                2._rt, *current_fp[lev][idim], 0,
+                0, 1, current_fp_temp[lev][idim]->nGrowVect()
+            );
+        }
+    }
+
+    // Calculate the electron pressure at t=n+1
+    CalculateElectronPressure(DtType::Full);
+
+    // Update the E field to t=n+1 using the extrapolated J_i^n+1 value
+    CalculateCurrentAmpere();
+    HybridPICSolveE(DtType::Full);
+
+    // Copy the rho^{n+1} values to rho_fp_temp and the J_i^{n+1/2} values to
+    // current_fp_temp since at the next step those values will be needed as
+    // rho^{n} and J_i^{n-1/2}.
+    for (int lev = 0; lev <= finest_level; ++lev)
+    {
+        MultiFab::Copy(*rho_fp_temp[lev], *rho_fp[lev],
+                        0, 0, 1, rho_fp_temp[lev]->nGrowVect());
+        for (int idim = 0; idim < 3; ++idim) {
+            MultiFab::Copy(*current_fp_temp[lev][idim], *current_fp[lev][idim],
+                           0, 0, 1, current_fp_temp[lev][idim]->nGrowVect());
+        }
+    }
+}
+
+void WarpX::CalculateCurrentAmpere ()
+{
+    for (int lev = 0; lev <= finest_level; ++lev)
+    {
+        CalculateCurrentAmpere(lev);
+    }
+}
+
+void WarpX::CalculateCurrentAmpere (int lev)
+{
+    WARPX_PROFILE("WarpX::CalculateCurrentAmpere()");
+
+    m_fdtd_solver_fp[lev]->CalculateCurrentAmpere(
+        current_fp_ampere[lev], Bfield_fp[lev],
+        m_edge_lengths[lev], lev
+    );
+
+    // we shouldn't apply the boundary condition to J since J = J_i - J_e but
+    // the boundary correction was already applied to J_i and the B-field
+    // boundary ensures that J itself complies with the boundary conditions, right?
+    // ApplyJfieldBoundary(lev, Jfield[0].get(), Jfield[1].get(), Jfield[2].get());
+    for (int i=0; i<3; i++) get_pointer_current_fp_ampere(lev, i)->FillBoundary(Geom(lev).periodicity());
+}
+
+void WarpX::HybridPICSolveE (DtType a_dt_type)
+{
+    for (int lev = 0; lev <= finest_level; ++lev)
+    {
+        HybridPICSolveE(lev, a_dt_type);
+    }
+    FillBoundaryE(guard_cells.ng_FieldSolver, WarpX::sync_nodal_points);
+}
+
+void WarpX::HybridPICSolveE (int lev, DtType a_dt_type)
+{
+    WARPX_PROFILE("WarpX::HybridPICSolveE()");
+
+    HybridPICSolveE(lev, PatchType::fine, a_dt_type);
+    if (lev > 0)
+    {
+        amrex::Abort(Utils::TextMsg::Err(
+        "HybridPICSolveE: Only one level implemented for hybrid-PIC solver."));
+        HybridPICSolveE(lev, PatchType::coarse, a_dt_type);
```

Comment:
I did leave it with the intent for it to be a hint as to what the function call would look like if mesh refinement is implemented for the hybrid solver at some point. But I agree we can actually remove it since it isn't a very profound hint.